kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs

```diff
@@ -0,0 +1,182 @@
+   Does not use cache, consider using 'cachedOIDCAuth'.
+-}
+oidcAuth :: DetectAuth
+oidcAuth AuthInfo{authProvider = Just(AuthProviderConfig "oidc" (Just cfg))} (tls, kubecfg)
+  = Just
+    $ parseOIDCAuthInfo cfg
+    >>= either (throwM . OIDCAuthParsingException) (\oidc -> pure (tls, addAuthMethod kubecfg oidc))
+oidcAuth _ _ = Nothing
+
+-- TODO: Consider doing this whole function atomically, as two threads may miss the cache simultaneously
+{-
+   Detects if auth-provier name is oidc, if it is configures the 'KubernetesClientConfig' with OIDCAuth 'AuthMethod'.
+   First looks for Auth information to be present in 'OIDCCache'. If found returns that, otherwise creates new Auth information and persists it in cache.
+-}
+cachedOIDCAuth :: OIDCCache -> DetectAuth
+cachedOIDCAuth cache AuthInfo{authProvider = Just(AuthProviderConfig "oidc" (Just cfg))} (tls, kubecfg) = Just $ do
+  m <- readTVarIO cache
+  o <- case findInCache m cfg of
+    Left e -> throwM $ OIDCAuthParsingException e
+    Right (Just o) -> return o
+    Right Nothing -> do
+      o@(OIDCAuth{..}) <- parseOIDCAuthInfo cfg
+                          >>= either (throwM . OIDCAuthParsingException) pure
+      let newCache = Map.insert (issuerURL, clientID) o m
+      _ <- atomically $ swapTVar cache newCache
+      return o
+  pure (tls, addAuthMethod kubecfg o)
+cachedOIDCAuth _ _ _ = Nothing
+
+findInCache :: Map (Text, Text) a -> Map Text Text -> Either String (Maybe a)
+findInCache cache cfg = do
+  issuerURL <- lookupEither cfg "idp-issuer-url"
+  clientID <- lookupEither cfg "client-id"
+  return $ Map.lookup (issuerURL, clientID) cache
+
+parseOIDCAuthInfo :: Map Text Text -> IO (Either String OIDCAuth)
+parseOIDCAuthInfo m = do
+  eitherTLSParams <- parseCA m
+  idTokenTVar <- atomically $ newTVar $ Map.lookup "id-token" m
+  refreshTokenTVar <- atomically $ newTVar $ Map.lookup "refresh-token" m
+  return $ do
+    tlsParams <- eitherTLSParams
+    issuerURL <- lookupEither m "idp-issuer-url"
+    clientID <- lookupEither m "client-id"
+    clientSecret <- lookupEither m "client-secret"
+    return OIDCAuth{..}
+
+parseCA :: Map Text Text -> IO (Either String TLS.ClientParams)
+parseCA m = do
+  t <- defaultTLSClientParams
+  fromMaybe (pure $ pure t) (parseCAFile t m <|> parseCAData t m)
+
+parseCAFile :: TLS.ClientParams -> Map Text Text -> Maybe (IO (Either String TLS.ClientParams))
+parseCAFile t m = do
+  caFile <- Text.unpack <$> Map.lookup "idp-certificate-authority" m
+  return $ updateClientParams t <$> BS.readFile caFile
+
+parseCAData :: TLS.ClientParams -> Map Text Text -> Maybe (IO (Either String TLS.ClientParams))
+parseCAData t m = do
+  caText <- Map.lookup "idp-certificate-authority-data" m
+  pure . pure
+    $ (B64.decode $ Text.encodeUtf8 caText)
+    >>= updateClientParams t
+
+lookupEither :: (Show key, Ord key) => Map key val -> key -> Either String val
```

Comment:
All of the `Either String` stuff here makes me a bit uncomfortable; it's the type of thing that would make this code difficult to work with when you wanted to handle and recover from specific errors.

A concrete error type (if there is only one error) or an error ADT (if there's more than one error that could occur together here) would be nicer if there's something that fits.

Reply:
This was very valuable, thanks for pointing it out. I have replcaed most of `Either String` with `Either <some exception>` and `throwM`ed the exception at the last possible moment. Let me know if looks good.